qualtran/bloqs/block_encoding/tensor_product.py

```diff
@@ -0,0 +1,154 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from functools import cached_property, reduce
+from typing import Dict, Sequence, Tuple
+
+from attrs import evolve, field, frozen, validators
+
+from qualtran import (
+    bloq_example,
+    BloqBuilder,
+    BloqDocSpec,
+    QAny,
+    QDType,
+    Register,
+    Signature,
+    SoquetT,
+)
+from qualtran.bloqs.block_encoding import BlockEncoding
+from qualtran.bloqs.block_encoding.lcu_select_and_prepare import PrepareOracle
+from qualtran.bloqs.bookkeeping import Partition
+from qualtran.symbolics import SymbolicFloat
+
+
+@frozen
+class TensorProduct(BlockEncoding):
+    r"""Tensor product of a sequence of block encodings.
+
+    Builds the block encoding as
+    $$
+        B[U_1 ⊗ U_2 ⊗ \cdots ⊗ U_n] = B[U_1] ⊗ B[U_2] ⊗ \cdots ⊗ B[U_n]
+    $$
+
+    When each $B[U_i]$ is a $(\alpha_i, a_i, \epsilon_i)$-block encoding of $U_i$, we have that
+    $B[U_1 ⊗ \cdots ⊗ U_n]$ is a $(\prod_i \alpha_i, \sum_i a_i, \sum_i \alpha_i \epsilon_i)$-block
+    encoding of $U_1 ⊗ \cdots ⊗ U_n$.
+
+    Args:
+        U: A sequence of block encodings.
+
+    Registers:
+        system: The system register.
+        ancilla: The ancilla register.
+        resource: The resource register.
+
+    References:
+        [Quantum algorithms: A survey of applications and end-to-end complexities](https://arxiv.org/abs/2310.03011). Dalzell et al. (2023). Ch. 10.2.
+    """
+
+    U: Sequence[BlockEncoding] = field(
```

Comment:
we've been moving to just using `Sequence` as the type annotation since it's more readable than tuple with its ellipses and it raises fewer false positives from type checkers that can't deduce types by respecting the `converter` argument.

Reply:
But that's the wrong type? Shouldn't class attributes have more specific types (similar to return values) and the constructor have generic types? Having the right type would also help us identify issues where we forget to provide a converter to convert the input to a tuple; which is important and would cause problems with hashing. 

Eg: https://mypy-play.net/?mypy=latest&python=3.12&gist=160e99b38930a8fcc5564ab7f790593b

Maybe @dstrain115 can chime in on advice for what's the best pattern.

Reply:
I think it is okay for a class method to be a Sequence, but then it should not be an error if there is a list or other sequence assigned to U.  (For instance, if you want to be flexible in the types that U accepts.  

If U has to be a tuple, then it should be a Tuple type.

Reply:
I changed the name to `block_encodings` and the type to `Tuple`.